lib/IDE/Refactoring.cpp

```diff
@@ -6576,7 +6609,27 @@ class AsyncConverter : private SourceEntityWalker {
     assert(TopHandler.getAsHandlerCall(const_cast<CallExpr *>(CE)) == CE &&
            "addHandlerCall must be used with a call to the TopHandler's "
            "completion handler");
-    auto Exprs = TopHandler.extractResultArgs(CE);
+    auto Exprs = TopHandler.extractResultArgs(CE, /*ReturnErrorArgsIfAmbiguous=*/true);
+    if (TopHandler.isAmbiguousCallToParamHandler(CE) && Exprs.isError()) {
+      // If the completion handler call is ambiguous, we use it as an error call
+      // by default. Apply heuristics to see if the error is already handled and
+      // whether the call should be interpreted as a success call.
+      assert(Exprs.args().size() == 1 && "There should only be one error parameter");
+      bool ErrorIsPlaceholered = false;
+      if (auto DRE = dyn_cast<DeclRefExpr>(Exprs.args().back())) {
+        if (Placeholders.count(DRE->getDecl())) {
+          ErrorIsPlaceholered = true;
+        }
+      }
+
+      if (ErrorIsPlaceholered) {
+        // The error parameter's decleration doesn't exist anymore. This is
+        // probably because it's been handled by a 'try await'. Since the
+        // original call was ambiguous, the result case seems more reasonable
+        // now.
+        Exprs = TopHandler.extractResultArgs(CE, /*ReturnErrorArgsIfAmbiguous=*/false);
+      }
```

Comment:
That does sound like a reasonable and simpler idea. 👍

Reply:
Do you think it's worth adding as a test case?

Reply:
We are actually by now converting this to code without placeholders because of all the various changes 🎉 Still, I added a test case because you can never have enough test cases.

```swift
func testReturnHandling4() async throws -> String {
  do {
    let str = try await simpleErr(arg: "xxx")
    return str
  } catch let err {
    print("some error stuff")
    throw err
  }
}
```

Reply:
Thanks!